File: js/predict.js
// ==================== 预测算法模块 ====================
// 包含：金银胆、五码、位置码、四注、和值、遗漏值

// 计算和值
function getSum(num1, num2, num3) {
    return parseInt(num1) + parseInt(num2) + parseInt(num3);
}

// 计算奇偶比
function getJiOu(nums) {
    let ji = 0;
    for (let n of nums) {
        if (parseInt(n) % 2 === 1) ji++;
    }
    return { ji, ou: 3 - ji };
}

// 计算大小比
function getDaXiao(nums) {
    let da = 0;
    for (let n of nums) {
        if (parseInt(n) >= 5) da++;
    }
    return { da, xiao: 3 - da };
}

// 分析遗漏值
function analyzeMiss(history) {
    const miss = {};
    for (let i = 0; i <= 9; i++) {
        miss[i] = 0;
    }

    for (let item of history) {
        const nums = [item.num1, item.num2, item.num3];
        for (let n of nums) {
            miss[parseInt(n)] = 0;
        }
        // 增加其他数字的遗漏值
        for (let i = 0; i <= 9; i++) {
            if (!nums.includes(String(i))) {
                miss[i]++;
            }
        }
    }
    return miss;
}

// 统计数字出现频率
function getNumFrequency(history, count = 30) {
    const freq = {0: 0, 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 7: 0, 8: 0, 9: 0};
    for (let i = 0; i < Math.min(count, history.length); i++) {
        const item = history[i];
        freq[parseInt(item.num1)]++;
        freq[parseInt(item.num2)]++;
        freq[parseInt(item.num3)]++;
    }
    return freq;
}

// 获取频率最高的N个数字
function getTopN(freq, n = 5) {
    return Object.entries(freq)
        .sort((a, b) => b[1] - a[1])
        .slice(0, n)
        .map(([k]) => parseInt(k));
}

// 获取位置频率
function getPositionFrequency(history, count = 50) {
    const pos1 = {0: 0, 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 7: 0, 8: 0, 9: 0};
    const pos2 = {0: 0, 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 7: 0, 8: 0, 9: 0};
    const pos3 = {0: 0, 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 7: 0, 8: 0, 9: 0};

    for (let i = 0; i < Math.min(count, history.length); i++) {
        const item = history[i];
        pos1[parseInt(item.num1)]++;
        pos2[parseInt(item.num2)]++;
        pos3[parseInt(item.num3)]++;
    }

    return { pos1, pos2, pos3 };
}

// 预测和值
function predictSum(history, count = 50) {
    const sumFreq = {};
    for (let i = 0; i < Math.min(count, history.length); i++) {
        const s = history[i].sum;
        sumFreq[s] = (sumFreq[s] || 0) + 1;
    }
    return Object.entries(sumFreq)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 2)
        .map(([k]) => parseInt(k));
}

// 生成四注推荐
function generateBets(wuxiao1, wuxiao2, wuxiao3, sums) {
    const bets = [];
    // 第一注：各取第一位
    bets.push([wuxiao1[0], wuxiao2[0], wuxiao3[0]].join(''));
    // 第二注：频率最高的组合
    bets.push([wuxiao1[0], wuxiao2[1], wuxiao3[1]].join(''));
    // 第三注：和值匹配
    for (let s of sums) {
        const target = s;
        for (let i = 0; i < wuxiao1.length; i++) {
            for (let j = 0; j < wuxiao2.length; j++) {
                for (let k = 0; k < wuxiao3.length; k++) {
                    if (wuxiao1[i] + wuxiao2[j] + wuxiao3[k] === target) {
                        bets.push([wuxiao1[i], wuxiao2[j], wuxiao3[k]].join(''));
                        return bets;
                    }
                }
            }
        }
    }
    // 默认第三注
    bets.push([wuxiao1[1], wuxiao2[0], wuxiao3[2]].join(''));
    return bets;
}

// 执行预测
function doPrediction(type) {
    const history = LotteryData[type];
    if (history.length === 0) {
        return null;
    }

    // 1. 金银胆
    const freq = getNumFrequency(history, 30);
    const sortedFreq = Object.entries(freq).sort((a, b) => b[1] - a[1]);
    const gold = parseInt(sortedFreq[0][0]);
    const silver = parseInt(sortedFreq[1][0]);

    // 2. 五码推荐
    const top5 = getTopN(freq, 5);
    const posFreq = getPositionFrequency(history, 50);
    const wuxiao = {
        pos1: getTopN(posFreq.pos1, 5),
        pos2: getTopN(posFreq.pos2, 5),
        pos3: getTopN(posFreq.pos3, 5)
    };

    // 3. 和值预测
    const sums = predictSum(history, 50);

    // 4. 四注推荐
    const bets = generateBets(wuxiao.pos1, wuxiao.pos2, wuxiao.pos3, sums);

    // 5. 334断组
    const duan = analyze334(history);

    // 6. 遗漏值
    const miss = analyzeMiss(history);

    return {
        gold,
        silver,
        top5,
        wuxiao,
        sums,
        bets,
        duan,
        miss,
        historyCount: history.length
    };
}

// 导出模块
window.analyzeMiss = analyzeMiss;
window.getNumFrequency = getNumFrequency;
window.getTopN = getTopN;
window.getPositionFrequency = getPositionFrequency;
window.predictSum = predictSum;
window.generateBets = generateBets;
window.doPrediction = doPrediction;

File: predict.py
import json
from collections import Counter

with open('c:/Users/zhao/WorkBuddy/Claw/data/all_history.json', 'r', encoding='utf-8') as f:
    data = json.load(f)

data_3d = sorted([d for d in data if d['type'] == '3d'], key=lambda x: x['period'], reverse=True)
data_pl3 = sorted([d for d in data if d['type'] == 'pl3'], key=lambda x: x['period'], reverse=True)

def get_num_frequency(history, count=30):
    """获取数字出现频率"""
    freq = {i: 0 for i in range(10)}
    for item in history[:count]:
        num = item['number']
        freq[int(num[0])] += 1
        freq[int(num[1])] += 1
        freq[int(num[2])] += 1
    return freq

def get_position_frequency(history, count=50):
    """获取位置频率"""
    pos1, pos2, pos3 = {i: 0 for i in range(10)}, {i: 0 for i in range(10)}, {i: 0 for i in range(10)}
    for item in history[:count]:
        num = item['number']
        pos1[int(num[0])] += 1
        pos2[int(num[1])] += 1
        pos3[int(num[2])] += 1
    return {'pos1': pos1, 'pos2': pos2, 'pos3': pos3}

def get_top_n(freq, n=5):
    """获取频率最高的N个数字"""
    return [int(k) for k, v in sorted(freq.items(), key=lambda x: x[1], reverse=True)[:n]]

def predict_sum(history, count=50):
    """预测和值"""
    sums = {i: 0 for i in range(28)}
    for item in history[:count]:
        num = item['number']
        s = int(num[0]) + int(num[1]) + int(num[2])
        sums[s] += 1
    # 取频率最高的2个和值
    return [int(k) for k, v in sorted(sums.items(), key=lambda x: x[1], reverse=True)[:2]]

def analyze_334(history, count=30):
    """334断组分析"""
    groups = {
        'g1': {'nums': [0, 1, 2], 'count': 0},
        'g2': {'nums': [3, 4, 5], 'count': 0},
        'g3': {'nums': [6, 7, 8, 9], 'count': 0}
    }
    for item in history[:count]:
        num = item['number']
        nums = [int(n) for n in num]
        for n in nums:
            if n in groups['g1']['nums']: groups['g1']['count'] += 1
            if n in groups['g2']['nums']: groups['g2']['count'] += 1
            if n in groups['g3']['nums']: groups['g3']['count'] += 1
    return groups

def analyze_55(history, count=30):
    """55分解分析"""
    group_a, group_b = 0, 0
    for item in history[:count]:
        num = item['number']
        nums = [int(n) for n in num]
        if any(n <= 4 for n in nums): group_a += 1
        if any(n >= 5 for n in nums): group_b += 1
    return {'a': group_a, 'b': group_b}

def generate_bets(pos1, pos2, pos3, sums):
    """生成四注推荐"""
    bets = []
    used = set()

    # 方法: 直接组合位置TOP5
    for i in range(5):
        for j in range(5):
            for k in range(5):
                bet = f'{pos1[i]}{pos2[j]}{pos3[k]}'
                if bet not in used:
                    used.add(bet)
                    bets.append(bet)
                if len(bets) >= 4:
                    break
            if len(bets) >= 4:
                break
        if len(bets) >= 4:
            break

    return bets[:4]

def analyze(data, name):
    print(f'\n========== {name} ==========')
    print(f'最新期号: {data[0]["period"]} ({data[0]["date"]})')
    print(f'最新号码: {data[0]["number"]}')
    print()

    # 1. 金银胆 (30期)
    freq30 = get_num_frequency(data, 30)
    sorted_freq = sorted(freq30.items(), key=lambda x: x[1], reverse=True)
    gold = sorted_freq[0][0]
    silver = sorted_freq[1][0]
    print(f'金胆: {gold} ({sorted_freq[0][1]}次/30期)')
    print(f'银胆: {silver} ({sorted_freq[1][1]}次/30期)')

    # 2. 五码 (30期频率)
    top5 = get_top_n(freq30, 5)
    print(f'五码: {" ".join(map(str, top5))}')

    # 3. 位置频率 (50期)
    pos_freq = get_position_frequency(data, 50)
    pos1_top5 = get_top_n(pos_freq['pos1'], 5)
    pos2_top5 = get_top_n(pos_freq['pos2'], 5)
    pos3_top5 = get_top_n(pos_freq['pos3'], 5)
    print(f'百位TOP5: {" ".join(map(str, pos1_top5))}')
    print(f'十位TOP5: {" ".join(map(str, pos2_top5))}')
    print(f'个位TOP5: {" ".join(map(str, pos3_top5))}')

    # 4. 和值 (50期)
    sum_pred = predict_sum(data, 50)
    print(f'和值推荐: {sum_pred[0]}, {sum_pred[1]}')

    # 5. 四注
    bets = generate_bets(pos1_top5, pos2_top5, pos3_top5, sum_pred)
    print(f'四注: {" ".join(bets)}')

    # 6. 334断组 (30期)
    duan = analyze_334(data, 30)
    print(f'334断组: 012={duan["g1"]["count"]}次 345={duan["g2"]["count"]}次 6789={duan["g3"]["count"]}次')
    weak = min(duan.items(), key=lambda x: x[1]['count'])[0]
    print(f'较弱组: {weak}')

    # 7. 55分解 (30期)
    fj = analyze_55(data, 30)
    print(f'55分解: 0-4={fj["a"]}次 5-9={fj["b"]}次')

    return {
        'gold': gold,
        'silver': silver,
        'top5': top5,
        'bets': bets,
        'sums': sum_pred
    }

# 执行预测
print('=' * 50)
print('2026-04-03 预测结果 (与网页算法一致)')
print('=' * 50)

result_3d = analyze(data_3d, '3D')
result_pl3 = analyze(data_pl3, '排列三')
